binding.cc

```diff
@@ -2,189 +2,147 @@
+  void *ptr = sodium_malloc(len);
+  assert(ptr != nullptr);
 
-  SN_STATUS_THROWS(js_detach_arraybuffer(env, array_buf), "failed to detach array buffer");
-
-  void *ptr;
-  err = js_remove_wrap(env, array_buf, &ptr);
+  js_value_t *buffer;
+  int err = js_create_external_arraybuffer(env, ptr, len, sn_external_arraybuffer_finalize, NULL, &buffer);
   assert(err == 0);
-  assert(ptr == buf_data);
 
-  sodium_free(buf_data);
-
-  int64_t ext_mem;
-  err = js_adjust_external_memory(env, -4 * 4096, &ext_mem);
-  assert(err == 0);
-  return NULL;
+  return static_cast<js_arraybuffer_t>(buffer);
 }
 
-js_value_t *
-sn_sodium_malloc (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(1, sodium_malloc);
-
-  SN_ARGV_UINT32(size, 0)
-
-  void *ptr = sodium_malloc(size);
-  SN_THROWS(ptr == NULL, "ENOMEM")
-
-  SN_THROWS(ptr == NULL, "sodium_malloc failed");
-
-  js_value_t *buffer;
-
-  SN_STATUS_THROWS(js_create_external_arraybuffer(env, ptr, size, NULL, NULL, &buffer), "failed to create a native arraybuffer")
-
+static void
+sn_sodium_free(js_env_t *env, js_receiver_t, js_arraybuffer_t buf) {
+  int err;
 
-  js_value_t *value;
-  SN_STATUS_THROWS(js_get_boolean(env, true, &value), "failed to create boolean")
-  SN_STATUS_THROWS(js_set_named_property(env, buffer, "secure", value), "failed to set secure property")
+  js_value_t *handle = static_cast<js_value_t *>(buf);
 
-  err = js_wrap(env, buffer, ptr, sn_sodium_free_finalise, NULL, NULL);
+  bool is_detached;
+  err = js_is_detached_arraybuffer(env, handle, &is_detached);
   assert(err == 0);
 
-  int64_t ext_mem;
-  err = js_adjust_external_memory(env, 4 * 4096, &ext_mem);
+  std::span<uint8_t> view;
+  err = js_get_arraybuffer_info(env, buf, view);
   assert(err == 0);
 
-  return buffer;
+  if (is_detached || view.data() == nullptr) return;
+
+  err = js_detach_arraybuffer(env, handle);
+  assert(err == 0);
 }
 
-js_value_t *
-sn_sodium_mprotect_noaccess (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(1, sodium_mprotect_noaccess);
+static inline int
+sn_sodium_mprotect_noaccess(js_env_t *env, js_receiver_t, js_arraybuffer_t buf) {
+  std::span<uint8_t> view;
 
-  SN_ARGV_TYPEDARRAY_PTR(buf, 0)
+  int err = js_get_arraybuffer_info(env, buf, view);
+  assert(err == 0);
 
-  SN_RETURN(sodium_mprotect_noaccess(buf_data), "failed to lock buffer")
+  return sodium_mprotect_noaccess(view.data());
 }
 
+static inline int
+sn_sodium_mprotect_readonly(js_env_t *env, js_receiver_t, js_arraybuffer_t buf) {
+  std::span<uint8_t> view;
 
-js_value_t *
-sn_sodium_mprotect_readonly (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(1, sodium_readonly);
-
-  SN_ARGV_TYPEDARRAY_PTR(buf, 0)
+  int err = js_get_arraybuffer_info(env, buf, view);
+  assert(err == 0);
 
-  SN_RETURN(sodium_mprotect_readonly(buf_data), "failed to unlock buffer")
+  return sodium_mprotect_readonly(view.data());
 }
 
+static inline int
+sn_sodium_mprotect_readwrite(js_env_t *env, js_receiver_t, js_arraybuffer_t buf) {
```

Comment:
This can use `js_arraybuffer_span_t` to avoid the explicit `js_get_arraybuffer_info()` call.